### ISSUE TITLE
MongoDb not serializing PropertyStorageProviders

### ISSUE THREAD

**Issue by stuartmcgillivray** [elsa-workflows/elsa-core#3103]
I noticed there was an issue with the property storage providers not returning the previously set value in the designer i.e. I want my ExampleActivity to store an ActivityOutput in blob storage and clicking save doesn't persist that property to the database.

I had a look into it and it seems the mongo db driver doesn't serialize readonly properties by default so PropertyStorageProviders in ActivityDefinition is by default left out and not saved to mongo.

https://github.com/elsa-workflows/elsa-core/blob/6daa8e456e8a3ab2b20da20d214626f259881b1f/src/core/Elsa.Abstractions/Models/ActivityDefinition.cs#L16

We can opt-in to storing this property by registering a class map, However this only appear to allow it to be serialized, not deserialised so it is still an empty list when it's read back.
```csharp
BsonClassMap.RegisterClassMap<ActivityDefinition>(cm =>
{
    cm.AutoMap();
    cm.MapProperty(c => c.PropertyStorageProviders);
});
```
A solution may be to just add the set in but I don't know if that would have implications in other places.

```csharp
public IDictionary<string, string> PropertyStorageProviders { get; set; } = new Dictionary<string, string>();
```

**Comment by terry-delph** [comment]
Would this registration fix the issue without any changes to the ActivityDefinition class?
```
BsonClassMap.RegisterClassMap<ActivityDefinition>(cm => 
{ 
     cm.AutoMap();
     new ImmutableTypeClassMapConvention().Apply(cm);
});
```

[ImmutableTypeClassMapConvention](https://github.com/mongodb/mongo-csharp-driver/blob/master/src/MongoDB.Bson/Serialization/Conventions/ImmutableTypeClassMapConvention.cs)

**Comment by stuartmcgillivray** [comment]
It doesn't, the ImmutableTypeClassMapConvention tries to use constructors to populate the read only values, the activity definition doesn't currently have a constructor.

**Comment by sfmskywalker** [comment]
Closed via #3150
